### PR TITLE
fix: fix oort utility calculation and add medmnist example for oort

### PR DESCRIPTION
**Pull request by jaemin-shin**
We change utility calculation of Oort to be done based on the first minibatch of the first local epoch, which was originally suggested by the Oort authors. We add new medmnist example for Oort, which includes graph that shows improved performance with Oort over Random selector.

## Description

Please provide a meaningful description of what this change will do, or is for. Bonus points for including links to
related issues, other PRs, or technical references.

Note that by _not_ including a description, you are asking reviewers to do extra work to understand the context of this
change, which may lead to your PR taking much longer to review, or result in it not being reviewed at all.

## Type of Change

- [x] Bug Fix
- [ ] New Feature
- [ ] Breaking Change
- [ ] Refactor
- [ ] Documentation
- [ ] Other (please describe)

## Checklist

- [ ] I have read the [contributing guidelines](/CONTRIBUTING.md)
- [ ] Existing issues have been referenced (where applicable)
- [ ] I have verified this change is not present in other open pull requests
- [ ] Functionality is documented
- [ ] All code style checks pass
- [ ] New code contribution is covered by automated tests
- [ ] All new and existing tests pass

### CODE REVIEW

File: lib/python/examples/medmnist_oort/trainer/main.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+        self.model.train()
+        self.model.to(self.device)
+
+        self.reset_stat_utility()
+
+        for epoch in range(1, self.epochs + 1):
+            self._train_epoch(epoch)
+        
+        self.model.to("cpu")
```
Comment (GustavBaumgart) [inline lib/python/examples/medmnist_oort/trainer/main.py:157]: Did this throw an error before this line was added?

File: lib/python/flame/mode/horizontal/oort/trainer.py [hunk 2]
```diff
@@ -80,17 +80,20 @@ def oort_loss(
...
-            loss_list = loss_fn(output, target, reduction="none")
+        if epoch == 1 and batch_idx == 0:
+            criterion = loss_fn(reduction='none')
+            loss_list = criterion(output, target)
             self._stat_utility += torch.square(loss_list).sum()
             loss = loss_list.mean()
         else:
-            loss = loss_fn(output, target)
+            criterion = loss_fn()
+            loss = criterion(output, target)
```
Comment (GustavBaumgart) [inline lib/python/flame/mode/horizontal/oort/trainer.py:96]: Is it possible that the loss function doesn't always have the parameter `reduction`?

File: lib/python/flame/selector/oort.py [hunk 3]
```diff
@@ -49,7 +49,7 @@ def __init__(self, **kwargs):
 
         if self.aggr_num < 0:
             self.aggr_num = 1
-        self.round = 0
+        self.round = -1
```
Comment (GustavBaumgart) [inline lib/python/flame/selector/oort.py:52]: Couldn't quite find out why this is -1

**Review by GustavBaumgart — COMMENTED** [review]
Good example; was nice to see the results! I left a few comments. Some of them only need clarification

File: lib/python/examples/medmnist_oort/aggregator/main.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+"""MedMNIST Oort aggregator for PyTorch."""
+
+import logging
+import numpy as np
+import torch
+import torchvision
+
+from flame.common.util import install_packages
+from flame.config import Config
+from flame.dataset import Dataset  # Not sure why we need this.
```
[the remark this answers is not in the document]
Comment (myungjin) [reply lib/python/examples/medmnist_oort/aggregator/main.py:25]: We can remove this. Here `Dataset` is used as type hint.

File: lib/python/examples/medmnist_oort/aggregator/main.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        return x
+
+
+class PyTorchMedMNistAggregator(TopAggregator):
+    """PyTorch MedMNist Aggregator"""
+
+    def __init__(self, config: Config) -> None:
+        self.config = config
+        self.model = None
+        self.dataset: Dataset = None  # Not sure why we need this.
```
[the remark this answers is not in the document]
Comment (myungjin) [reply lib/python/examples/medmnist_oort/aggregator/main.py:93]: We can simple do `self.dataset = None`. Right now, dataset is declared as a required attribute in TopAggregator class.

We can remove the declaration (https://github.com/cisco-open/flame/blob/main/lib/python/flame/mode/horizontal/syncfl/top_aggregator.py#L65-L71) from the base class (TopAggregator).
We can do that in a separate PR.

File: lib/python/flame/selector/oort.py (hunk 3 above)
Comment (myungjin) [reply lib/python/flame/selector/oort.py:52]: Yes, I have the same question. Please add some comment in the file if necessary.

**Review by myungjin — APPROVED** [review]
left a few minor questions. overall looks good.

**Review by myungjin — APPROVED** [review]
left a few minor questions. overall looks good.

File: lib/python/flame/selector/oort.py (hunk 3 above)
Comment (jaemin-shin) [reply lib/python/flame/selector/oort.py:52]: This is because channel.property["round"] is 0 at first round and 2 at second round; 

because of this, trainer selection was not happening due to line 88-89, which is made to select the same set of ends at a same round when select() is called multiple times.

I think I should keep this line as self.round = 0, and change other parts instead as follows:
1. Change line 88-89 if statement

if round <= self.round and len(self.selected_ends) != 0:
    return {key: None for key in self.selected_ends}

2. Set channel.property["round"] to be 1 at first round (could be a separate PR)

File: lib/python/examples/medmnist_oort/README.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+```
+
+Once this is complete, and say, you want to run the example (using all 10 trainers), you can select a number of trainers K to aggregate per round, and run `python run.py <K>`.
+If you want to run the example with aggregating 3 trainers, you can run:
+
+```bash
+python run.py 3
+```
+
+Note that Oort selects ```1.3 * K``` at a round, and aggregates ```K``` trainers in order of their model updates' arrival.
```
[the remark this answers is not in the document]
Comment (jaemin-shin) [reply lib/python/examples/medmnist_oort/README.md]: Done

File: lib/python/examples/medmnist_oort/aggregator/main.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jaemin-shin) [reply lib/python/examples/medmnist_oort/aggregator/main.py]: Agreed! Done

File: lib/python/examples/medmnist_oort/aggregator/main.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    parser = argparse.ArgumentParser(description='')
+    parser.add_argument('config', nargs='?', default="./config.json")
+
+    args = parser.parse_args()
+
+    config = Config(args.config)
+
+    a = PyTorchMedMNistAggregator(config)
+    a.compose()
+    a.run()
```
[the remark this answers is not in the document]
Comment (jaemin-shin) [reply lib/python/examples/medmnist_oort/aggregator/main.py]: Done

File: lib/python/examples/medmnist_oort/config/agg_template.json [hunk 8]
```diff
@@ -40,23 +40,23 @@
...
+        "rounds": 100,
+        "epochs": 4
     },
     "baseModel": {
         "name": "",
-        "version": 2
+        "version": 1
     },
     "job": {
         "id": "622a358619ab59012eabeefb",
```
[the remark this answers is not in the document]
Comment (jaemin-shin) [reply lib/python/examples/medmnist_oort/config/agg_template.json]: Was not intentional, simply using the same thing from other examples. Changed the job id for now.

File: lib/python/examples/medmnist_oort/trainer/main.py (hunk 1 above)
Comment (jaemin-shin) [reply lib/python/examples/medmnist_oort/trainer/main.py:162]: Haven't checked, but this example uses CUDA if it is available so I think we need this line?

File: lib/python/flame/mode/horizontal/oort/trainer.py (hunk 2 above)
Comment (jaemin-shin) [reply lib/python/flame/mode/horizontal/oort/trainer.py:96]: Good point. I added lines in trainer/main.py which checks if the loss function has that parameter. I initially thought of putting that line in this function, but that would require too much overhead to do it every time the loss function is called

**Review by myungjin — COMMENTED** [review]
lgtm

**Review by myungjin — APPROVED** [review]

**Review by GustavBaumgart — APPROVED** [review]
LGTM
